### PR TITLE
Leave types validations

### LINKED ISSUE #1700: Add validation on leave type settings
### Description
When we create a leave type, there are no validation on allocation_period and  allocation_frequency.

https://github.com/saeloun/miru-web/assets/10629762/36a7b89a-0f37-4149-bf17-5d8a6a32881b

### What is the expected behaviour?
If allocation_period and allocation_frequency are invalid then the leave type should not be saved.
Instead we should show the appropriate error.

### How to reproduce the issue?

1.
2.
3.

### PR DESCRIPTION
**Pull request by apoorv1316**
Fixes #1700

### CODE REVIEW

File: app/models/leave_type.rb [hunk 1]
```diff
@@ -84,4 +86,37 @@ class LeaveType < ApplicationRecord
...
+          "Invalid combination: Allocation period in months can only have frequency per quarter or per year")
+      end
+    end
+
+    def valid_allocation_value
+      max_values = {
+        ["days", "per_week"] => 7,
+        ["days", "per_month"] => 31,
+        ["days", "per_quarter"] => 93,
+        ["days", "per_year"] => 365,
```
Comment (nisusam) [inline app/models/leave_type.rb:107]: `366` for leap year

File: app/models/leave_type.rb [hunk 2]
```diff
@@ -84,4 +86,37 @@ class LeaveType < ApplicationRecord
...
+          :base,
+          "Invalid combination: Allocation period in months can only have frequency per quarter or per year")
+      end
+    end
+
+    def valid_allocation_value
+      max_values = {
+        ["days", "per_week"] => 7,
+        ["days", "per_month"] => 31,
+        ["days", "per_quarter"] => 93,
```
Comment (nisusam) [inline app/models/leave_type.rb:106]: `92`

**Review by nisusam — CHANGES_REQUESTED** [review]

File: app/models/leave_type.rb [hunk 3]
```diff
@@ -84,4 +87,74 @@ class LeaveType < ApplicationRecord
     def discard_timeoff_entries
       timeoff_entries.discard_all
     end
+
+    def valid_allocation_combination
```
Comment (prasanthchaduvula) [inline app/models/leave_type.rb:91]: If possible move these validations to concerns

File: app/models/concerns/leave_type_validatable.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    end
+
+    def valid_allocation_value
+      max_values = {
+        ["days", "per_week"] => 7,
+        ["days", "per_month"] => 31,
+        ["days", "per_quarter"] => 92,
+        ["days", "per_year"] => 366,
+        ["weeks", "per_month"] => 5,
+        ["weeks", "per_quarter"] => 13,
```
Comment (nisusam) [inline app/models/concerns/leave_type_validatable.rb:33]: 92 / 7 = 13.14

File: app/models/concerns/leave_type_validatable.rb [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+        ["weeks", "per_year"] => 52,
```
Comment (nisusam) [inline app/models/concerns/leave_type_validatable.rb:34]: 366 / 7 = 52.28

**Review by nisusam — APPROVED** [review]
Just rebase with `develop`

**Pull request merged**